overlord/snapstate/snapstate.go

```diff
@@ -1349,16 +1361,36 @@ func UpdateUnderDeviceContext(st *state.State, name string, opts *RevisionOption
 			SideInfo:    snapst.CurrentSideInfo(),
 			Flags:       snapst.Flags.ForSnapSetup(),
 			InstanceKey: snapst.InstanceKey,
-		}
+			CohortKey:   opts.CohortKey,
+		}
+
+		if snapst.Channel != opts.Channel || snapst.CohortKey != opts.CohortKey {
+			var summary string
+			switch {
+			case snapst.Channel == opts.Channel:
+				summary = fmt.Sprintf(i18n.G("Switch snap %q from cohort %q to %q"), snapsup.InstanceName(),
+					strutil.ElliptRight(snapst.CohortKey, 10),
+					strutil.ElliptRight(opts.CohortKey, 10),
+				)
+			case snapst.CohortKey == opts.CohortKey:
+				summary = fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q"), snapsup.InstanceName(),
+					snapst.Channel, opts.Channel)
+			default:
+				summary = fmt.Sprintf(i18n.G("Switch snap %q from channel %q to %q and from cohort %q to %q"), snapsup.InstanceName(),
+					snapst.Channel, opts.Channel,
+					strutil.ElliptRight(snapst.CohortKey, 10),
+					strutil.ElliptRight(opts.CohortKey, 10),
+				)
+			}
 
-		if snapst.Channel != opts.Channel {
-			// update the tracked channel
+			// update the tracked channel and cohort
 			snapsup.Channel = opts.Channel
+			snapsup.CohortKey = opts.CohortKey
 			// Update the current snap channel as well. This ensures that
 			// the UI displays the right values.
 			snapsup.SideInfo.Channel = opts.Channel
 
-			switchSnap := st.NewTask("switch-snap-channel", fmt.Sprintf(i18n.G("Switch snap %q from %s to %s"), snapsup.InstanceName(), snapst.Channel, opts.Channel))
+			switchSnap := st.NewTask("switch-snap-channel", summary)
```

Comment:
Do we test the summary? It seems to make sense now that it gets constructed in a more "complicated" way.